### PR TITLE
Attributes metadata

### PR DESCRIPTION
**Pull request by zdne**
This pull requests introduces arbitrary metadata on the grouping level. 

Note in a future this can be done – in theory – on any level. At this moment 
the metadata are plain textual (string) key-value pairs. 

The `meta` attribute is a hash of metadata from different sources. The sources 
can be either user (author – writer – of the source file) or adapter (that 
"loaded" the source file into API refract). 

The "adapter" metadata are currently not used and reserved for future use of 
adapters. For example adapter version or information about the source file. 

The "user" metadata can be used to hold any metadata as specified in the source
file by the user. For example consider following blueprint: 

``` apib
HOST: http://my.api.into
FORMAT: 1A
Author: Z
Email: z@apiary.io

# My API
```

Note the adapter – parser – SHOULD NOT have semantic understanding to the 
"user" metadata.

Remarks & improvements welcomed!

### CODE REVIEW

File: namespaces/api-namespace.md [hunk 1]
```diff
@@ -125,6 +140,15 @@ Given an API description with following layout:
 }
 ```
 
+## String Pair (object)
```
Comment (danielgtaylor) [inline namespaces/api-namespace.md]: If we are introducing this new type, should we also use it for HTTP headers?

**Comment by danielgtaylor** [comment]
Generally I think this looks good. I'm a little worried that a newcomer would get confused at the use of both `element.meta` and `element.attributes.meta` but I don't have a better suggestion for the name.

File: namespaces/api-namespace.md [hunk 2]
```diff
@@ -42,6 +42,11 @@ For example a `category` element may be classified both as `resourceGroup` and
             - resourceGroup - Category is a set of resource.
             - dataStructures - Category is a set of data structures.
             - scenario - Reserved. Category is set of steps.
+- `attributes`
+    - `meta` (object) - Arbitrary metadata
```
Comment (fosrias) [inline namespaces/api-namespace.md]: I think for the sake of clarity, this should be `metadata` here. It is something completely different than `meta` and the extra verbosity will keep the confusion down.

Comment (smizell) [inline namespaces/api-namespace.md]: A couple of things:
1. I'm not sure I see that big of a difference myself between `meta` and `metadata`
2. ~~Since we've been using camel case, shouldn't it be `metaData`?~~ Looks like it can be used as one word
3. In Minim, we will be providing convenience properties for meta attributes, meaning you'll use `element.id` and `element.title` for accessing `element.meta.id` and so on. This may shield the user enough from this issue.

Comment (smizell) [inline namespaces/api-namespace.md]: Edit, `metadata` can be one word.

File: namespaces/api-namespace.md [hunk 3]
```diff
@@ -42,6 +42,11 @@ For example a `category` element may be classified both as `resourceGroup` and
...
+        - `user` (array[String Pair]) - User-specific metadata. Metadata written in the source.
+        - `adapter` (array[String Pair]) - Serialization-specific metadata. Metadata provided by adapter.
```
Comment (smizell) [inline namespaces/api-namespace.md]: Are you meaning to use these `String Pair` objects this way, where they look like the content of a `Member Element`?

Comment (smizell) [inline namespaces/api-namespace.md]: In other words, `Member Element` are elements have a content for key/value, whereas this is simply an object.

Comment (smizell) [inline namespaces/api-namespace.md]: Gosh, jet lag mind I guess... I see now you are not using key/value as elements. Please strike these comments from the record.

**Comment by smizell** [comment]
@zdne Makes sense. I have some concern with departing from Refract and making this complex structure in attributes. Seems like you could create a `meta` element and provide the same data with being extendible in the future. Just a thought.

File: namespaces/api-namespace.md (hunk 3 above)
Comment (zdne) [inline namespaces/api-namespace.md]: @smizell I was thinking about using the `member` element but I think it is way too complex for what i need at this moment. It can fallback into elements in a future.

**Comment by zdne** [comment]
> Seems like you could create a meta element and provide the same data with being extendible in the future.

Where would meta live? I am willing to set on any solution. Just give me a proposal. This PR is just one of the possible solution for what I need to address (capture arbitrary user defined metadata from API description documents as well as any adapter-specific metadata). 

Please let me know what to do to get this PR merged :smile:  It is not clear from your comments.

File: namespaces/api-namespace.md (hunk 2 above)
Comment (pksunkara) [inline namespaces/api-namespace.md]: I agree with renaming it to `metadata` to avoid confusion.

File: namespaces/api-namespace.md (hunk 1 above)
Comment (pksunkara) [inline namespaces/api-namespace.md]: A key in header can be used multiple times, so we need to be careful in deciding if we want to use this.

**Comment by pksunkara** [comment]
IMHO I don't like the `String Pair`. We are moving from normal AST to Refract AST, then the key-value pairs should also be somehow represented in refract and should be able to use refract's advantages. Using `String Pair` is like mixing AST with Refract AST which is a worse since it's like mixing both of them.

**Comment by smizell** [comment]
Just some ideas. My purpose is to provide some ideas to drive some discussion. This is what I mean when I say use Refract instead of complex structures (such as `meta` property that is a primitive array of primitive objects). I would prefer to see something along these lines.

## Idea 1: Use Member Elements

Use current Member Element.

``` json
{
  "element": "category",
  "attributes": {
    "meta": [
      {
        "element": "member",
        "meta": {
          "class": [ "user" ]
        },
        "content": {
          "key": {
            "element": "string",
            "content": "HOST"
          },
          "value": {
            "element": "string",
            "content": "http://my.api.into"
          }
        }
      }
    ]
  }
}
```

## Idea 2: Special Meta Element Inside Attributes

Use an object as the content of `meta`. This is OK within the Refract spec. I also added a `head` property instead to make it more generic. This can be used in any of these examples, and is just an idea, not a hard proposal.

``` json
{
  "element": "category",
  "attributes": {
    "head": [
      {
        "element": "meta",
        "meta": {
          "class": [ "user" ]
        },
        "content": {
          "key": "HOST",
          "value": "http://my.api.into"
        }
      }
    ]
  }
}
```

## Idea 3: Special Meta Subclassed

There would also be an `adapterMeta` element. Similar to above, just not using class.

``` json
{
  "element": "category",
  "attributes": {
    "head": [
      {
        "element": "userMeta",
        "content": {
          "key": "HOST",
          "value": "http://my.api.into"
        }
      }
    ]
  }
}
```

## Idea 4: Meta in Content

Note: Really, any of the above examples could be put into content like this.

``` json
{
  "element": "category",
  "content": [
    {
      "element": "userMeta",
      "content": {
        "key": "HOST",
        "value": "http://my.api.into"
      }
    }
  ]
}
```

**Comment by pksunkara** [comment]
Is this

``` json
"content": {
  "key": {
    "element": "string",
    "content": "HOST"
  },
  "value": {
    "element": "string",
    "content": "http://my.api.into"
  }
}
```

equivalent to this

``` json
"content": {
    "key": "HOST",
    "value": "http://my.api.into"
}
```

in refract?

If yes, I vote for 4, otherwise I vote for 1.

**Comment by smizell** [comment]
@pksunkara Your examples are currently not equivalent, though we experimented with making them equivalent. We could discuss that at another point. 

The one issue with the first example is that it is not a Refract object, but rather just an array of member elements. What we have done in the resource namespace is repeat the property as an element. Here, the `metadata` element would extend an MSON object.

``` json
{
  "element": "category",
  "attributes": {
    "metadata": {
      "element": "metadata",
      "content": [
        {
          "element": "member",
          "meta": {
            "class": [ "user" ]
          },
          "content": {
            "key": {
              "element": "string",
              "content": "HOST"
            },
            "value": {
              "element": "string",
              "content": "http://my.api.into"
            }
          }
        }
      ]
    }
  }
}
```

I think it may look a lot better if we can use member elements like I've shown in example one for hashes like this. The example in this comment and the [Href Variables](https://github.com/refractproject/refract-spec/blob/master/namespaces/resource-namespace.md#href-variables-object-type) example seems a little verbose.

@zdne I would like to see some discussion around whether item #1 is a better/worse option than using normal objects as you've outlined. If it's not better, we can move forward with your proposal.

File: namespaces/api-namespace.md (hunk 1 above)
Comment (zdne) [inline namespaces/api-namespace.md]: Do not be too much focused on this type – it is merely a helper to keep things dry. Note it is not even an element. As for HTTP headers – I want them to use MSON syntax / refract eventually.

**Comment by zdne** [comment]
@smizell let's do no.1 – I had it initially there but found it too complex for what was needed. I am going to update the PR.

**Comment by zdne** [comment]
@smizell updated

**Comment by smizell** [comment]
@zdne One unanswered question: do you want to use `metadata` instead of `meta` as brought up by @fosrias?

https://github.com/refractproject/refract-spec/pull/30#discussion_r32538481

**Comment by zdne** [comment]
> One unanswered question: do you want to use metadata instead of meta as brought up by

I consider this odd. Semantically both are meta or metadata one related to refract element definition while the other are related to the actual content of the element. So to me there is no difference between `meta` and `metadata`. Sure one is present on refract level and the other on attributes level but then we can call them `meta1` and `meta2` with the same success. 

Another thing would be call those `refractMeta` and `contentMeta` ... 

TL;DR: I would just stay with `meta` (or `contentMeta` if you really want to dis-ambiguate these two)...

**Comment by pksunkara** [comment]
I have a new opinion now. Implementing the no.1 was a bit tough and needs some workarounds in C++ drafter.
